src/dbn_struct.jl

```diff
@@ -0,0 +1,174 @@
+
+Reference to one or more named nodes in the POMDP or MDP dynamic Bayesian network (DBN).
+
+Note that `gen(::DBNOut, m, s, a, rng)` always takes `s` and `a` arguments (the inputs to the entire DBN) while `gen(::DBNVar, m, depargs..., rng)` takes a variable number of arguments (one for each dependency).
+
+`DBNOut` is a "value type". See the documentation of `Val` for more conceptual details about value types.
+"""
+struct DBNOut{names} end
+
+DBNOut(name::Symbol) = DBNOut{name}()
+DBNOut(names...) = DBNOut{names}()
+
+struct DBNDef{N<:NamedTuple}
+    nodes::N
+    deps::NamedTuple # this could be a parameter in the future
+end
+
+# todo DBNDef structure
+
+node(d::DBNDef, name::Symbol) = d.nodes[name]
+deps(d::DBNDef, name::Symbol) = d.deps[name]
+nodenames(d::DBNDef) = keys(d.nodes)
+@generated function add_node(d::DBNDef, n::DBNVar{name}, node, deps) where name
+    quote 
+        @assert !haskey(d.nodes, name) "DBNDef already has a node named :$name"
+        DBNDef(merge(d.nodes, ($name=node,)), merge(d.deps, ($name=deps,)))
+    end
+end
+
+function mdp_dbn()
+    DBNDef((s = InputDBNNode(),
+            a = InputDBNNode(),
+            sp = DistributionDBNNode(transition),
+            r = FunctionDBNNode(reward),
+           ),
+           (s = (),
+            a = (),
+            sp = (:s, :a),
+            r = (:s, :a, :sp),
+           )
+          )
+end
+
+function pomdp_dbn()
+    DBNDef((s = InputDBNNode(),
+            a = InputDBNNode(),
+            sp = DistributionDBNNode(transition),
+            o = DistributionDBNNode(observation),
+            r = FunctionDBNNode(reward),
+           ),
+           (s = (),
+            a = (),
+            sp = (:s, :a),
+            o = (:s, :a, :sp),
+            r = (:s, :a, :sp, :o),
+           )
+          )
+end
+
+"""
+    DBNStructure(::Type{M}) where M <: Union{MDP, POMDP}
+
+Trait of an MDP/POMDP type for describing the structure of the dynamic Baysian network.
+
+# Example
+
+    struct MyPOMDP <: MDP{Int, Int} end
+    POMDPs.gen(::MyPOMDP, s, a, rng) = (sp=s+a+rand(rng, [1,2,3]), r=s^2)
+
+    # make a new node delta_s that is deterministically sp-s
+    POMDPs.DBNStructure(::Type{MyPOMDP}) = add_node(mdp_dbn(),
+                                                    DVNVar(:delta_s),
+                                                    FunctionDBNNode((s,sp)->sp-s),
+                                                    (:s, :sp))
+    gen(DBNOut(:delta_s), MyPOMDP(), 1, 1, Random.GLOBAL_RNG)
+"""
+function DBNStructure end
+
+DBNStructure(::Type{M}) where M <: MDP = mdp_dbn()
+DBNStructure(::Type{M}) where M <: POMDP = pomdp_dbn()
+
+DBNStructure(m) = DBNStructure(typeof(m))
+
+struct InputDBNNode end # this does nothing for now
+
+struct DistributionDBNNode{F}
+    dist_func::F
+end
+
+@generated function gen(n::DistributionDBNNode, m, args...)
+    # apparently needs to be @generated for type stability
+    argexpr = (:(args[$i]) for i in 1:length(args)-1)
```

Comment:
just out of curiosity can you explain?

Reply:
Yeah, if you just do `n.dist_func(m, args[1:end-1]...)`, the compiler doesn't know how to infer the expansion of `args[1:end-1]` and so can't figure out which method of dist_func to call. If you do this expansion manually at compile time in the @generated part, then it can infer everything correctly at compile time.